### ISSUE TITLE
Running code in VSCode

### ISSUE THREAD

**Issue by KarinaKarina6** [cmu-phil/py-tetrad#16]
Hello!
I would like to know about ways to run codes. From the README I understood that files can be launched through the terminal, and this method works for me. But I would like to work with files through Visual Studio Code to be able to debug the code. Is it possible to do this? I tried to run file `run_continuous.py ` through Visual Studio Code, but I encountered the following error, although everything works fine when launched from the terminal.
![image](https://github.com/cmu-phil/py-tetrad/assets/55412063/d907377a-99fd-4864-8b83-811be0488ce8)

**Comment by jdramsey** [comment]
Let me think.... I'm used to doing this in PyCharm; let me think of how to do it in VsCode...

**Comment by jdramsey** [comment]
Hmm... I can't see the whole file. Do you have something like this in your file?

import jpype.imports

try:
    jpype.startJVM(classpath=[f"resources/tetrad-current.jar"])
except OSError:
    print("JVM already started")

**Comment by jdramsey** [comment]
(Well, I mean, with the correct indentation...)

**Comment by KarinaKarina6** [comment]
Sorry, maybe I'm doing something wrong.

I opened the `run_continuous.py` file and ran the code step by step. This line works without errors:

`import jpype.imports`

After this the code is executed:

`jpype.startJVM(classpath=[f"resources/tetrad-current.jar"])`

Next is done:

`import pandas as pd`

But when I get to the line:

`import tools.translate as tr`

The error that I sent you in the screenshot appears.

**Comment by KarinaKarina6** [comment]
As far as I understand, there is `translate.py` code in file:

```
try:
     jpype.startJVM(classpath=[f"resources/tetrad-current.jar"])
except OSError:
     pass
```

`except OSError` occurs

**Comment by jdramsey** [comment]
OK, tomorrow sometime, I'll pull up VSCode and give it a shot. Are you on Mac? Windows? I have both... I don't have Linux...

**Comment by KarinaKarina6** [comment]
I'm on Windows. Thank you!

**Comment by jdramsey** [comment]
Also, let me open up my Windows laptop and see if I can get it work in VSCode... wish me luck--I don't usually use Windows...

**Comment by KarinaKarina6** [comment]
I understand, I wish you good luck! I hope it won't be a big problem

**Comment by jdramsey** [comment]
Sorry for the delay; I got caught up in some other projects. Later today and tomorrow I'm going to spend some time on my Windows laptop (which I don't do much) so I'll have a look at this again.

**Comment by cdecker8** [comment]
I've been wrestling with the same issue in VSCode, specifically regarding imports from the .edu.cmu namespace. Running in a  Linux VM container with Python 3.9 and JPype1==1.3.0. So it's not an OS specific issue. 

An interesting observation is that importing TetradSearch.py works seamlessly using the syntax import tools.TetradSearch as ts. So far, I've managed to complete most tasks by utilizing pandas for data preprocessing, initializing a search object, and invoking functions through the search object.

The challenge arises when attempting to directly access any of the .edu.cmu codebase. Notably, when inspecting files within the tools directory, such as TetradSearch.py, VSCode indicates a lack of access to those imports. Interestingly, this restriction seems to vanish when importing that specific file into another script, where it appears to function correctly— throwing no input errors and is able to output plausible graphs.

My investigation into this issue has been limited thus far, but it seems that jpype, the module facilitating Java integration, might be encountering difficulties in passing back classes from the Java Virtual Machine (JVM). I haven't found a solution other than only working with the imported python objects and not trying to call any of the java codebase directly.

**Comment by jdramsey** [comment]
Huh, maybe that's the issue I'm having with it. I'll try it again this weekend and see if that's what's holding it up for me. Thanks.

**Comment by cdecker8** [comment]
Yeah, I'm not entirely sure what's going on. It's somewhat comforting to know that someone else is facing a similar issue because I've been in full-on PhD imposter mode, thinking I must be making some glaring mistake. Well, I haven't ruled out the possibility of me doing something wrong, but...

For what it's worth, I did confirm that my container is running a 64-bit version of Python and OpenJDK. If I manage to figure out what's going on in my container with the .edu imports I'll update with more info.

**Comment by jdramsey** [comment]
Thanks!

**Comment by cdecker8** [comment]
I have mine running in VScode after playing around with paths some. It really doesn't like relative paths if you are running in VSCode.

Here's the version of continuous.py that works for me. Obviously the BASE DIR is specific to my set up and needs to be updated.

`import jpype.imports
**import sys

BASE_DIR = "/workspace/notebooks/Causal/py-tetrad/pytetrad/"
sys.path.append(BASE_DIR)**

try:
    **jpype.startJVM(classpath=[f"{BASE_DIR}/**resources/tetrad-current.jar"])
except OSError:
    print("JVM already started")

import pandas as pd

import tools.translate as tr
import tools.TetradSearch as ts

**data = pd.read_csv(f"{BASE_DIR}resources/**airfoil-self-noise.continuous.txt", sep="\t")
data = data.astype({col: "float64" for col in data.columns})

## Make a TetradSearch instance to run searches against. This helps to organize
## the use of Tetrad search algorithms and hides the JPype code for those who
## don't want to deal with it.
search = ts.TetradSearch(data)
search.set_verbose(False)

## Pick the score to use, in this case a continuous linear, Gaussian score.
search.use_sem_bic()
search.use_fisher_z(alpha=0.05)

## Run various algorithms and print their results. For now (for compability with R)
print('FGES')
search.run_fges()
print(search.get_graph_to_matrix())

print('BOSS')
search.run_boss(num_starts=1, use_bes=True, time_lag=0, use_data_order=True)
print(search.get_string())
dag=search.get_dag_java()
print(dag)

search.run_boss()
adj = search.get_graph_to_matrix()

print(adj)

print('SP')
search.run_sp()
print(search.get_string())

print('GRaSP')
search.run_grasp()
print(search.get_string())

print('PC')
search.run_pc()
print(search.get_string())

print('FCI')
search.run_fci()
print(search.get_string())

print('CFCI')
search.run_cfci()
print(search.get_string())

print('GFCI')
search.run_gfci()
print(search.get_string())

print('BFCI')
search.run_bfci()
print(search.get_string())

print('GRaSP-FCI')
search.run_grasp_fci()
print(search.get_string())

print('CCD')
search.run_ccd()
print(search.get_string())

print('SVAR-GFCI')
search.run_svar_gfci()
print(search.get_string())

print('SP-FCI')
search.run_spfci()
print(search.get_string())

print('ICA-LiNGAM')
search.run_ica_lingam(threshold_b=0.4, threshold_spine=0)
print(search.get_string())

print('ICA-LiNG-D')
search.run_ica_lingd(threshold_b=0.4, threshold_spine=0)
print(search.get_string())


`

**Comment by cdecker8** [comment]
I have mine running. It appears to be an issue with using relative paths as changing to use direct paths seems to fix the problem. Here's my working version of continuous.py

import jpype.imports
import sys

BASE_DIR = "/workspace/notebooks/Causal/py-tetrad/pytetrad/"
sys.path.append(BASE_DIR)

try:
    jpype.startJVM(classpath=[f"{BASE_DIR}/resources/tetrad-current.jar"])
except OSError:
    print("JVM already started")

import pandas as pd

import tools.translate as tr
import tools.TetradSearch as ts

data = pd.read_csv(f"{BASE_DIR}resources/airfoil-self-noise.continuous.txt", sep="\t")
data = data.astype({col: "float64" for col in data.columns})

The rest is unchanged

**Comment by jdramsey** [comment]
Thanks so much! I'll try this.

**Comment by jdramsey** [comment]
By the way, in case Karina hasn't seen this, I'll tag har: @KarinaKarina6

**Comment by KarinaKarina6** [comment]
@cdecker8 thanks a lot! I'll try this soon. I hope that everything will work out, because I really want to work with this package.
